stores/sql/chain.go

```diff
@@ -118,26 +119,141 @@ func UpdateWalletSiacoinElementProofs(ctx context.Context, tx sql.Tx, updater wa
+		el, err := scanSiacoinStateElement(rows)
+		if err != nil {
+			return nil, fmt.Errorf("failed to scan state element: %w", err)
+		}
+		elements = append(elements, el)
+	}
+	return elements, rows.Err()
+}
+
+func updateSiacoinStateElements(ctx context.Context, tx sql.Tx, elements []SiacoinStateElement) error {
+	updateStmt, err := tx.Prepare(ctx, "UPDATE wallet_outputs SET leaf_index = ?, merkle_proof= ? WHERE output_id = ?")
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to update state elements: %w", err)
+	}
+	defer updateStmt.Close()
+
+	for _, el := range elements {
+		if _, err := updateStmt.Exec(ctx, el.LeafIndex, MerkleProof{el.MerkleProof}, el.ID); err != nil {
+			return fmt.Errorf("failed to update state element '%v': %w", el.ID, err)
+		}
+	}
+	return nil
+}
+
+func FileContractElement(ctx context.Context, tx sql.Tx, fcid types.FileContractID) (types.V2FileContractElement, error) {
+	var contract V2Contract
+	var leafIndex uint64
+	var proof MerkleProof
+	err := tx.QueryRow(ctx, "SELECT contract, leaf_index, merkle_proof FROM contract_elements ce INNER JOIN contracts c ON ce.db_contract_id = c.id WHERE c.fcid = ?", Hash256(fcid)).
+		Scan(&contract, &leafIndex, &proof)
+	if err != nil {
+		return types.V2FileContractElement{}, err
+	}
+	return types.V2FileContractElement{
+		ID: fcid,
+		StateElement: types.StateElement{
+			LeafIndex:   leafIndex,
+			MerkleProof: proof.Hashes,
+		},
+		V2FileContract: types.V2FileContract(contract),
+	}, nil
+}
+
+func InsertFileContractElements(ctx context.Context, tx sql.Tx, fces []types.V2FileContractElement) error {
+	contractIDStmt, err := tx.Prepare(ctx, "SELECT c.id FROM contracts c WHERE c.fcid = ?")
+	if err != nil {
+		return err
+	}
+	defer contractIDStmt.Close()
+
+	insertStmt, err := tx.Prepare(ctx, "INSERT INTO contract_elements (created_at, db_contract_id, contract, leaf_index, merkle_proof) VALUES (?, ?, ?, ?, ?)")
+	if err != nil {
+		return err
+	}
+	defer insertStmt.Close()
+
+	for _, fce := range fces {
+		var contractID int64
+		err = contractIDStmt.QueryRow(ctx, Hash256(fce.ID)).Scan(&contractID)
+		if errors.Is(err, dsql.ErrNoRows) {
+			return contractNotFoundErr(fce.ID)
+		} else if err != nil {
+			return err
+		}
+		_, err = insertStmt.Exec(ctx, time.Now().UTC(), contractID, V2Contract(fce.V2FileContract), fce.StateElement.LeafIndex, MerkleProof{fce.StateElement.MerkleProof})
+		if err != nil {
+			return fmt.Errorf("failed to insert file contract element: %w", err)
+		}
+	}
+	return nil
+}
+
+func RemoveFileContractElements(ctx context.Context, tx sql.Tx, fcids []types.FileContractID) error {
+	contractIDStmt, err := tx.Prepare(ctx, "SELECT c.id FROM contracts c WHERE c.fcid = ?")
+	if err != nil {
+		return err
+	}
+	defer contractIDStmt.Close()
+
+	deleteStmt, err := tx.Prepare(ctx, "DELETE FROM contract_elements WHERE db_contract_id = ?")
+	if err != nil {
+		return err
+	}
+	defer deleteStmt.Close()
+
+	for _, fcid := range fcids {
```

Comment:
is this really necessary? I think you can argue it's safe to do `DELETE FROM WHERE id IN (SELECT id FROM ` here..

Reply:
Theoretically we could probably change it but that means building the placeholder string which isn't nice either. This feels a bit more readable and obvious. We also don't need to worry about performance here.